homeassistant/components/rainbird/switch.py

```diff
@@ -2,61 +2,96 @@
 
 import logging
 
+from pyrainbird import AvailableStations, RainbirdController
 import voluptuous as vol
 
-from homeassistant.components.switch import PLATFORM_SCHEMA, SwitchDevice
-from homeassistant.const import (
-    CONF_FRIENDLY_NAME,
-    CONF_SCAN_INTERVAL,
-    CONF_SWITCHES,
-    CONF_TRIGGER_TIME,
-    CONF_ZONE,
-)
+from homeassistant.components.switch import SwitchDevice
+from homeassistant.const import ATTR_ENTITY_ID, CONF_FRIENDLY_NAME, CONF_TRIGGER_TIME
 from homeassistant.helpers import config_validation as cv
 
-from . import DATA_RAINBIRD
+from . import DATA_RAINBIRD, DOMAIN, RAINBIRD_CONTROLLER
 
-DOMAIN = "rainbird"
 _LOGGER = logging.getLogger(__name__)
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+ATTR_DURATION = "duration"
+
+SERVICE_START_IRRIGATION = "start_irrigation"
+
+SERVICE_SCHEMA_IRRIGATION = vol.Schema(
     {
-        vol.Required(CONF_SWITCHES, default={}): vol.Schema(
-            {
-                cv.string: {
-                    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
-                    vol.Required(CONF_ZONE): cv.string,
-                    vol.Required(CONF_TRIGGER_TIME): cv.string,
-                    vol.Optional(CONF_SCAN_INTERVAL): cv.string,
-                }
-            }
-        )
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_DURATION): vol.All(vol.Coerce(float), vol.Range(min=0)),
     }
 )
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up Rain Bird switches over a Rain Bird controller."""
-    controller = hass.data[DATA_RAINBIRD]
 
+    if discovery_info is None:
+        return
+
+    controller: RainbirdController = hass.data[DATA_RAINBIRD][
+        discovery_info[RAINBIRD_CONTROLLER]
+    ]
+    available_stations: AvailableStations = controller.get_available_stations()
+    if not (available_stations and available_stations.stations):
+        return
     devices = []
-    for dev_id, switch in config.get(CONF_SWITCHES).items():
-        devices.append(RainBirdSwitch(controller, switch, dev_id))
+    for zone in range(1, available_stations.stations.count + 1):
+        if available_stations.stations.active(zone):
+            zone_config = discovery_info.get("zones", {}).get(zone, {})
+            time = zone_config.get(
+                CONF_TRIGGER_TIME, discovery_info.get(CONF_TRIGGER_TIME, 0)
+            )
+            name = zone_config.get(CONF_FRIENDLY_NAME)
+            if not time:
```

Comment:
Instead of checking here we should validate already in the config schema and set default for each controller based on the main setting. See the manual alarm_control_panel schema for example:
https://github.com/home-assistant/home-assistant/blob/770eeaf82f4800127e977f980f5438d0fa8edbc9/homeassistant/components/manual/alarm_control_panel.py#L63-L75

Reply:
this way?